### DISCUSSION TITLE
macos test failures

### DISCUSSION THREAD

**Discussion by eleith** [eleith/emailjs#311] [questions]
@zackschuster we are getting macos test failures on [latest merge](https://github.com/eleith/emailjs/actions/runs/2164161800). do they pass for you locally (if you use a mac)?

**Comment by zackschuster** [comment]
yup, on both intel & apple silicon. i think the failures we're seeing are from a combination of slowness in `mailparser` (based on profiling) and the sheer number of targets we're running in a commodity environment, which i would unscientifically rate as being at best on par with my 2015 macbook pro. the easy fix would be removing `node@10` support in `v4`, but really i just think we need to find a way to speed up our tests. as an example, even on an M1 Pro `message.ts › streams message` takes five seconds to complete.

**Reply by zackschuster** [reply]
#312 isn't showing any flakiness (yet), which is promising. it's still pretty slow but i think reducing the number of reads in `test/message.ts` & removing the `test-cjs` task is helping.

**Reply by eleith** [reply]
v4.0.0 still fails on macos for a few node versions. fyi.

**Reply by zackschuster** [reply]
😬 😬 😬 

i'll keep looking. i hadn't noticed this before, but _only_ macOS is this slow. linux & windows targets all finished `test/message.ts` in ~2s.

**Reply by zackschuster** [reply]
as of https://github.com/eleith/emailjs/commit/5eea9d0b884cd6f4b107caa93266d2ec264e50ff on my dev branch the tests are now completely non-blocking & complete in roughly half the time (on macOS; ubuntu & windows are slightly slower). they also still throw timeouts for the slowest tests in ci. 🤷🏻 🤷🏻 🤷🏻 

as the workflow logs show, the only bottleneck is `mailparser`, with the usual ~aggravating~ caveat that it's on macOS only. at this point i don't think there's much (if anything) more to be done to our code to improve the situation.

**Reply by zackschuster** [reply]
i've [opened an issue](https://github.com/nodemailer/mailparser/issues/317) on the `mailparser` repo with my findings.

**Reply by eleith** [reply]
thanks @zackschuster, what do you think about mocking mailparser fullly and reducing the scope of the test to just internal code?

**Reply by zackschuster** [reply]
by mocking, do you mean rolling our own parser? to my mind, the value of using `mailparser` is that we have a trusted community module verifying that we're transmitting the data correctly; otherwise, we're essentially saying "trust us, we know what we're doing". i would be fine with replacing it with something more performant, but npm searches don't lead me to believe we even really have a second choice.

**Comment by eleith** [comment]
no. I meant only in the unit test, replacing it with a test double

**Reply by zackschuster** [reply]
yeah, that was my assumption (i don't think the library has any business shipping an email parser). the same concerns apply.

**Reply by eleith** [reply]
I just looked at the test. I see this module is used explicitly to test aspects of the library, so a double makes no sense, it's not used otherwise.

unfortunate dependency I guess.

tests are passing now, so I assume we just have flakiness? we could disable macOS if we can't get it green in the future...

**Reply by zackschuster** [reply]
> I assume we just have flakiness?

yeah, specifically for the larger message tests on macOS. everything else is fine.

> we could disable macOS if we can't get it green in the future...

i think at this point ci is _just barely_ too slow on the flaky tests, so we basically need the stack or the environment to give us just about _any_ consistent perf win & we'll be solidly green.
